src/vivarium/framework/results.py

```diff
@@ -0,0 +1,325 @@
+from collections import defaultdict, Counter
+import itertools
+import typing
+from typing import Any, Callable, Dict, List, Tuple
+
+import pandas as pd
+
+if typing.TYPE_CHECKING:
+    from vivarium.framework.engine import Builder
+    from vivarium.framework.event import Event
+    from vivarium.framework.time import Time, Timedelta
+
+
+class ResultsContext:
+
+    def __init__(self):
+        # Binners split continuous population columns into categorical ones
+        self._binners = []
+        # Mappers apply a mapping function from a population column to a new column
+        self._mappers = []
+        # keys are event names
+        # values are dicts with key (filter, grouper) value (measure, aggregator, additional_keys)
+        self._producers = defaultdict(lambda: defaultdict(list))
+
+        self.population = None
+
+    def add_binner(self, target_col: str, result_col: str, bins: List, labels: List[str], **cut_kwargs):
+        """Adds a specification to bin a target column into a new column.
+
+        We frequently want to group results by bins of a continuous population
+        attribute.  For example, simulants might have an age and we'd
+        want to stratify results by 5-year age bins.  You could then
+        provide a binner that will preprocess the population by binning
+        the `age` attribute in the state table into an `age_group` attribute.
+        This is not stored in the state table, but dynamically created when
+        results are observed.
+
+        Parameters
+        ----------
+        target_col
+            The continuous column we want to bin into a categorical column.
+        result_col
+            The name of the resulting categorical column.
+        bins
+            The bin edges.
+        labels
+            The labels of the resulting bins.
+        cut_kwargs
+            Additional kwargs to provide to :func:`pandas.cut`.
+
+        """
+        # TODO: Some validation
```

Comment:
to include picking up duplication of result_col in all of these state table manipulations ?